rgsync/Connectors/redis_connector.py

```diff
@@ -56,81 +100,108 @@ def WriteData(self, data):
+            self.session = None             #Next time we will connect to database
             self.exactlyOnceLastId = None
             self.shouldCompareId = True if self.exactlyOnceTableName is not None else False
-            msg = "Failed to connect to Redis database, error = %s" % str(e)
+            msg = "Exception occur while getting shard id for exactly once. Exception : {}".format(e)
             WriteBehindLog(msg)
             raise Exception(msg) from None
 
-
+        pipe = self.session.pipeline()
         try:
 
             # data is list of dictionaries
             # iterate over one by one and extract a dictionary and process it
+            batch = []
+            isAddBatch = True if data[0]['value'][OP_KEY] == OPERATION_UPDATE_REPLICATE else False
+            lastStreamId = None
             for d in data:
                 d_val = d['value']              # get value of key 'value' from dictionary
 
-                lastStreamId = d.pop('id', None)    # pop the stream id out of the record, we do not need it
+                lastStreamId = d.pop('id', None)    # pop the stream id out of the record
                 if self.shouldCompareId and CompareIds(self.exactlyOnceLastId, lastStreamId) >= 0:
-                    WriteBehindLog('Skip %s as it was already writen to the backend' % lastStreamId)
+                    WriteBehindLog('Skip {} as it was already written to the backend'.format(lastStreamId))
                     continue
 
                 # check operation permission, what gets replicated
-                op = d_val.pop(OP_KEY, None)
+                op = d_val.pop(OP_KEY, None)        # pop the operation key out of the record 
                 if op not in self.supportedOperations:
                     msg = 'Got unknown operation'
                     WriteBehindLog(msg)
                     raise Exception(msg) from None
 
-                self.shouldCompareId = False
                 pk = d_val.pop(SIMPLE_HASH_BACKEND_PK)
-                newKey = '%s:%s' % (self.new_prefix, pk)
-
-                if op == OPERATION_UPDATE_REPLICATE:
-                    d = [[k, v] for k,v in d_val.items() if not k.startswith('_')]
-                    res = self.session.execute_command('HSET', newKey, *sum(d, []))
-                    if 'ERR' in str(res):
-                        WriteBehindLog("Error: Command HSET execution failed, error = %s" % str(res))
-                        raise Exception(res) from None 
+                newKey = '{}:{}'.format(self.new_prefix, pk)
+
+                self.shouldCompareId = False
+                if op != OPERATION_UPDATE_REPLICATE:
+                    if isAddBatch:
+                        # process update batch
+                        for kv_pair in batch:
+                            for k, v in kv_pair.items():
+                                pipe.hset(k, mapping=v)
+                        
+                        batch = []
+                        isAddBatch = False
```

Comment:
I believe you took it from the sql connector, on redis you do not really need it because you can do delete and hset on the same pipe.

Reply:
@MeirShpilraien 
yes I have tried to make it like sql I will upate the code in a week.
I have marked this as draft version